worldSetup.py

```diff
@@ -46,47 +47,78 @@ def getEntrancePool(self, settings, connectorsOnly=False):
-            if settings.entranceshuffle in ("expert", "insanity"):
+            if settings.shufflejunk:
                 types.update(["dummy", "trade"])
-            if settings.entranceshuffle in ("insanity",):
+            if settings.shuffleannoying:
                 types.add("insanity")
+            if settings.shufflewater:
+                types.add("water")
             if settings.randomstartlocation:
                 types.add("start")
             if settings.dungeonshuffle:
                 types.add("dungeon")
+            if settings.entranceshuffle in ("mixed"):
+                types.add("connector")
             entrances = [k for k, v in ENTRANCE_INFO.items() if v.type in types]
 
         return entrances
 
-    def randomize(self, settings, rnd):
+    def pickEntrances(self, settings, rnd):
         if settings.overworld == "dungeondive":
             self.entrance_mapping = {"d%d" % (n): "d%d" % (n) for n in range(9)}
         if settings.randomstartlocation and settings.entranceshuffle == "none":
             start_location = start_locations[rnd.randrange(len(start_locations))]
             if start_location != "start_house":
                 self.entrance_mapping[start_location] = "start_house"
                 self.entrance_mapping["start_house"] = start_location
-        
-        entrances = self.getEntrancePool(settings)
-        for entrance in entrances.copy():
-            self.entrance_mapping[entrance] = entrances.pop(rnd.randrange(len(entrances)))
 
-        # Shuffle connectors among themselves
-        entrances = self.getEntrancePool(settings, connectorsOnly=True)
-        for entrance in entrances.copy():
-            self.entrance_mapping[entrance] = entrances.pop(rnd.randrange(len(entrances)))
+        entrancePool = set(self.getEntrancePool(settings))
+        unmappedEntrances = list(entrancePool)
+
+        # Exiting from a closed D7 gets you stuck in the wall, don't let a connector come out there
+        if 'd7' in entrancePool and settings.entranceshuffle == 'mixed':
+            self.entrance_mapping['d7'] = rnd.choice([x for x in entrancePool if ENTRANCE_INFO[self.entrance_mapping[x]].type != 'connector'])
+            entrancePool.remove('d7')
+
+        for entrance in entrancePool:
+            self.entrance_mapping[entrance] = unmappedEntrances.pop(rnd.randrange(len(unmappedEntrances)))
+
+        if settings.entranceshuffle == 'split':
+            # Shuffle connectors among themselves
+            # entrancePool is intentionally overwritten so we're only swapping connectors
+            entrancePool = self.getEntrancePool(settings, connectorsOnly=True)
+            unmappedEntrances = list(entrancePool)
+
+            for entrance in entrancePool.copy():
+                self.entrance_mapping[entrance] = unmappedEntrances.pop(rnd.randrange(len(unmappedEntrances)))
+
+        # Make sure all entrances in the pool are accessible
+        for j in range(1000):
```

Comment:
If we reach the end of a 1000 and not found a solution, we better throw an exception here. Instead of later during item placement.